caffe2/operators/tensorrt_op_trt.cc

```diff
@@ -0,0 +1,184 @@
+      return false;
+    }
+  }
+  return true;
+}
+
+} // namespace
+
+// Upon construction, we build the inference enigne by deserializing from
+// protobuf string. And since we know the input/output blobs, we can do the
+// binding here too.
+TensorRTOp::TensorRTOp(const OperatorDef& operator_def, Workspace* ws)
+    : Operator<CUDAContext>(operator_def, ws),
+      logger_((nvinfer1::ILogger::Severity)(
+          OperatorBase::GetSingleArgument<int>("log_verbosity", 2))),
+      batch_size_(OperatorBase::GetSingleArgument<int>("max_batch_size", 1)) {
+  {
+    auto engine_string =
+        OperatorBase::GetSingleArgument<std::string>("serialized_engine", "");
+    CAFFE_ENFORCE(!engine_string.empty(), "Empty serialized TensorRT engine!");
+    auto trt_runtime = InferObject(nvinfer1::createInferRuntime(logger_));
+    // TODO(support trt plugin factory)
+    trt_engine_ = InferObject(trt_runtime->deserializeCudaEngine(
+        engine_string.data(), engine_string.size(), nullptr));
+  }
+
+  if(!trt_engine_) {
+    CAFFE_THROW("Cannot deserialize TensorRT engine!");
+  }
+
+  // match and bind the input/output
+  std::unordered_map<std::string, int> inputs;
+  std::unordered_map<std::string, int> outputs;
+  for (int i = 0; i < operator_def.input_size(); ++i) {
+    inputs.emplace(operator_def.input(i), i);
+    std::cout << "Adding Input: " << operator_def.input(i) << std::endl;
+  }
+  for (int i = 0; i < operator_def.output_size(); ++i) {
+    outputs.emplace(operator_def.output(i), i);
+    std::cout << "Adding Output: " << operator_def.output(i) << std::endl;
+  }
+
+
+  int num_bindings = trt_engine_->getNbBindings();
+  for (int b = 0; b < num_bindings; ++b) {
+    const auto& name = trt_engine_->getBindingName(b);
+    nv_dims_.push_back(trt_engine_->getBindingDimensions(b));
+    if (trt_engine_->bindingIsInput(b)) {
+      std::cout << "Checking TRT input: " << name << std::endl;
+      const auto it = inputs.find(name);
+      CAFFE_ENFORCE(it != inputs.end(), MakeString("Cannot find trt input: ", name));
+      binding_hints_.emplace_back(it->second, true);
+    } else {
+      const auto it = outputs.find(name);
+      CAFFE_ENFORCE(it != outputs.end());
+      binding_hints_.emplace_back(it->second, false);
+    }
+  }
+
+  trt_executor_ = InferObject(trt_engine_->createExecutionContext());
+}
+
+bool TensorRTOp::RunOnDevice() {
+  CAFFE_ENFORCE(trt_executor_);
+  // Decide input batch size
+  int N = 0;
+  bool first = true;
+  for (int i = 0; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    if (first) {
+      N = tensor_dims.front();
+      first = false;
+    } else {
+      CAFFE_ENFORCE(
+          N == tensor_dims.front(), "Mismatched batch size in input tensors");
+    }
+  }
+  CAFFE_ENFORCE(N <= batch_size_, "Batch size is too large");
+  batch_size_ = N;
+
+  // We need to do the binding at RunOnDevice time because we only know the
```

Comment:
it's usually a good idea to cache bindings between invocations in case pointers or sizes don't change. But it can be a TODO for now

Reply:
I thought about it but I wasn't sure if the sizes are guaranteed not to change or not. Binding itself doens't cost much. More computation is used in dim size comparison.